packages/react-native/ReactCommon/react/performance/timeline/PerformanceObserver.cpp

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include "PerformanceObserver.h"
+#include "PerformanceObserverRegistry.h"
+#include "PerformanceEntryReporter.h"
+
+namespace facebook::react {
+
+PerformanceObserver::~PerformanceObserver() {
+  if (auto registry = registry_.lock()) {
+    registry->removeObserver(*this);
+  }
+}
+
+void PerformanceObserver::pushEntry(const PerformanceEntry& entry) {
+  buffer_.add(entry);
+}
+
+std::vector<PerformanceEntry> PerformanceObserver::takeRecords() {
+  return buffer_.consume();
+}
+
+bool PerformanceObserver::isObserving(PerformanceEntryType type) const {
+  return observedTypes_.contains(type);
+}
+
+void PerformanceObserver::observe(PerformanceEntryType type, bool buffered) {
+  // we assume that `type` was checked on JS side and is correct
+  observedTypes_.clear();
```

Comment:
AFAIK this would be an error because single `type` cannot be changed.

Reply:
I mean it should never be executed if JS side validates this anyway

Reply:
I just realized this is incorrect. You can call `observe` with `type` multiple times to observe different types. E.g.:

```
performance.observe({type: 'mark', buffered: true});
performance.observe({type: 'measure', buffered: true});
```